src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,329 @@
+        if (!isset($options['database'])) {
+            throw new InvalidArgumentException(
+                'You must provide the "database" option for MongoDBStore'
+            );
+        }
+
+        $this->mongo = $mongo;
+
+        $this->options = array_merge(array(
+            'collection' => 'lock',
+        ), $options);
+
+        $this->initialTtl = $initialTtl;
+    }
+
+    /**
+     * Create a TTL index to automatically remove expired locks.
+     *
+     * This should be called once during database setup.
+     *
+     * Alternatively the TTL index can be created manually:
+     *
+     *  db.lock.ensureIndex(
+     *      { "expires_at": 1 },
+     *      { "expireAfterSeconds": 0 }
+     *  )
+     *
+     * Please note, expires_at is based on the application server. If the
+     * database time differs; a lock could be cleaned up before it has expired.
+     * Set a positive expireAfterSeconds to account for any time drift between
+     * application and database server.
+     *
+     * A TTL index MUST BE used on MongoDB 2.2+ to automatically clean up expired locks.
+     *
+     * @see http://docs.mongodb.org/manual/tutorial/expire-data/
+     *
+     * @return string The name of the created index
+     *
+     * @throws UnsupportedException          if options are not supported by the selected server
+     * @throws MongoInvalidArgumentException for parameter/option parsing errors
+     * @throws DriverRuntimeException        for other driver errors (e.g. connection errors)
+     */
+    public function createTTLIndex(): string
+    {
+        $keys = array(
+            'expires_at' => 1,
+        );
+
+        $options = array(
+            'expireAfterSeconds' => 0,
+        );
+
+        return $this->getCollection()->createIndex($keys, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws LockStorageException if failed to save to storage (fallback exception)
+     * @throws LockExpiredException when save is called on an expired lock
+     */
+    public function save(Key $key)
+    {
+        $token = $this->getUniqueToken($key);
+        $now = microtime(true);
+
+        $filter = array(
+            '_id' => (string) $key,
+            '$or' => array(
+                array(
+                    'token' => $token,
+                ),
+                array(
+                    'expires_at' => array(
+                        '$lte' => $this->createDateTime($now),
+                    ),
+                ),
+            ),
+        );
+
+        $update = array(
+            '$set' => array(
+                '_id' => (string) $key,
+                'token' => $token,
+                'expires_at' => $this->createDateTime($now + $this->initialTtl),
+            ),
+        );
+
+        $options = array(
+            'upsert' => true,
+        );
+
+        $key->reduceLifetime($this->initialTtl);
+
+        try {
+            $this->getCollection()->updateOne($filter, $update, $options);
```

Comment:
Yes, if for some reason the user doesn't create a TTL Index (auto deleting index). A lock that gets left behind (expired) can be picked up and re-used for a different thread. The (`or token equals`)  behavior is to allow successive calls to save to just work. For the later, see `AbstractStoreTest::testSaveTwice`, this is the intended behavior of all lock stores.

Reply:
Okay - that makes sense. Can we make the implementation such that if users don't call the `ttl` we set a default index which can be overridden with what user specifies? We don't want to leave the possibility of being able to leave documents behind in a collection and let it grow unknowingly.

Reply:
Please don't get mixed up between:

a lock ttl i.e. `public function __construct(Client $mongo, array $options, float $initialTtl = 300.0)`

and a MongoDb TTL Index i.e. https://docs.mongodb.com/manual/core/index-ttl/

A MongoDb TTL Index is a database index which causes documents to be DELETED by the DBMS. Adding a MongoDB TTL index is a Database administration task. The MongoDbStore should NOT do this. It does however provide a method `MongoDbStore::createTtlIndex` allowing a programmer to automate the creation of this index in their deployment workflow. I personally would never use this method as it should only be called once when the database is created.

Reply:
Right, I was't confused between the two, sorry for using the wrong terminology there :)

Reply:
so when you say:

> if users don't call the `ttl` we set a default index 

Are you talking about us automatically calling `MongoDbStore::createTtlIndex`?

Because this won't work on mongodb prior to version `2.2`. Also we can't assume the application servers and DB servers are clock synced. Creating a TTL index when they are out can cause locks to be removed before they expire. In the worst case scenario, a database server could be on a different timezone to an application server.

If the user is running their locks across a stack spanning multiple data centers in different geographic regions they may decide to set a relatively high `expireAfterSeconds`. Someone running everything in 1 data center and running NTP might decide to set `expireAfterSeconds = 0`